content/advent-2018/how-to-send-and-receive-sms.md

```diff
@@ -0,0 +1,501 @@
+
+In this post you'll learn what [UCP](https://wiki.wireshark.org/UCP) is and how we can use it to directly communicate with an [SMSC](https://en.wikipedia.org/wiki/Short_Message_service_center) to send and receive [SMS](https://en.wikipedia.org/wiki/SMS) messages using Go.
+
+## Terminology
+
+### Mobile Terminating Message
+
+Messages from telco to subscriber. For example, weather update messages.
+
+![mobile-terminating](/postimages/advent-2018/mobile-terminating.png)
+
+
+### Mobile Originating Message
+
+Messages from subscriber to telco. For example, texting a keyword to an accesscode for balance inquiry.
+
+![mobile-originating](/postimages/advent-2018/mobile-originating.png)
+
+### Multi-part MTs and MOs
+
+SMS greater than 160 characters are considered multi-part SMS.
+To send a multi-part mobile terminating messages, we need to split it into message parts. Each message part will contain a message part number, total message parts number, and a reference number. 
+
+The multi-part mobile originating message also contains a message part number, total message parts number, and a reference number. We need to concatenate those message parts in order to interpret the mobile originating message that the user sent.
+
+
+
+## Universal Computer Protocol
+
+
+UCP is primarily used to connect to short message service centres (SMSCs), in order to send and receive short messages (SMS).
+
+![protocol](/postimages/advent-2018/emiucp.svg)
+
+### session management operation
+
+Allows us the send login credentials to the SMSC.
+
+### alert operation
+
+Allows us to send pings to the SMSC.
+
+### submit short message operation
+
+Allows us to send mobile terminating SMS.
+
+### delivery notification operation
+
+Sent by the SMSC to the client as a delivery status receipt indicating if the SMS message was successfully sent or not.
+
+### delivery short message operation
+
+Sent by the SMSC to the client on behalf of a subscriber's mobile originating SMS message.
+
+## Implementation
+
+We can treat UCP like a traditional client-server protocol. After etablishing a TCP connection, we send UCP requests with a sequence number(called transaction reference number in the spec) from 00 to 99 and the SMSC will respond with a UCP response message synchronously. However the SMSC can also send UCP requests, just like in the case of `delivery notification operation` and `delivery short message operation`. We also need to periodically send keepalive pings to the SMSC so that it wont treat the connection as stale and disconnect us. You can read the protocol specification [here](http://documents.swisscom.com/product/1000174-Internet/Documents/Landingpage-Mobile-Mehrwertdienste/UCP_R4.7.pdf).
+
+
+
+Lets start with a `Client` struct containing the login credentials to the SMSC.
```

Comment:
In the "Conclusion" section you're saying 

> If you work on the telco field and have an access to an SMSC, feel free to try [..]

I feel it will be nice to add more context about the project we're going to build. As an example, the `Client` below has address and credentials. But how one can get them? I think a paragraph that describes the requirements for the `Client` will make the post more compleat.

Reply:
I'll add a section `Demo` wherein I'll use an SMSC simulator. I'll explore the UCP packets in Wireshark too. I'll finish up the simulator first, then add the github repo link so that the readers can follow along even though they wont send any real SMS.